### DISCUSSION TITLE
A Job processed by multiple machines simultaneously?

### DISCUSSION THREAD

**Discussion by rlawnsdudwo** [google/or-tools#2714] [CP-SAT questions] [answered]
Hello, I'm working with "flexible_job_shop_sat.py", also I'm python newbie :(
It seem like A job must be worked with only one machine at a time.
However, I'd like to change that A job can be occupied with multiple machine at a time.
For easy to understand, I'd like to group machines as team.
And also, I want to differentiate machine type.

**For example:**
Task 0 require 3 basic machines occupied to start.
Task 1 require 2 basic machines + 1 advanced machine occupied to start.
Task 2 require 1 basic machine + 2 advanced machines occupied to start.

+ all tasks have predecessor and successor relationship. (Task 0 → Task 1 → Task 2)
+ If there are only 1 basic machine, 1 advanced machine left, no jobs can be started.
+ if there are only 1 basic machine, 2 advanced machines left, but Task 1 is not done yet, Task 2 cannot be started.

Please help me!

**Comment by lperron** [comment]
not an issue. Moving to discussions.

**Reply by rlawnsdudwo** [reply]
Sorry, I am new to github. Thank you!

**Comment by gregy4** [comment]
Generally it is possible to model it with optional intervals which give you complete information which machines are used by a task.
For every task model an interval and set of optional intervals representing all possible machines.
Add equality constraint for basic and advanced machines to be equal with a number of corresponding active optional intervals.
Add noOverlap constraint for optional intervals over a machine.
Add enforcement constraint to guarantee that active optional interval is equal to task interval.

In case you are only interested in task positions and basic and advanced machines don't overlap, it is possible to model a problem with cumulative constraint per machine type. In that case optional intervals are not needed.

**Reply by rlawnsdudwo** [reply]
Thank you for answer. It sounds like I have to code about Intervals and Constraints. Will there be good interval and constraint coding example for me to study? Since I am still learning about coding, it is difficult for me to hard code yet.
I will search for it myself but it's hard to find about my case(still learning...).

Thank you very much.

**Reply by gregy4** [reply]
Understand. Flexible_job_shop_sat is good example, it is nearly what you want with optional intervals.

model.Add(sum(l_presences) == 1) is exact place where is defined number of used machines by a task. You should create two groups of l_presences (one for basic, another for advanced machines) and replace 1 with requested number of basic (or advanced machines) so you have two constraints that guarantee that needed number of machines is used.

**Reply by rlawnsdudwo** [reply]
Thank you for the guidance. I will try.
I have one more question, dont i have to adjust first part? Where i have to set up the task and machines?

**Reply by gregy4** [reply]
Yes, you should add relations between tasks. When task1 should by after task2, task1.end <= task2.start.
Machines actually doesn't exist, the point is in grouping of relevant intervals in noOverlap constraint to guarantee that a machine is not used twice at same time.

Validate solver results. From my experience it is good way to be sure that your model is correct or to identify a part that is not correctly modeled.

**Comment by lperron** [comment, marked as answer]
To answer you initial question, nothing prevent you from using the same interval in multiple no_overlap or cumulative constraints.
We do this in the rcpsp_sat.py example.
